### ISSUE TITLE
task cloning failed

### ISSUE THREAD

**Issue by kenneth59715** [cianfrocco-lab/COSMIC-CryoEM-Gateway#113]
log error:
Oct 27 2018 06:47:08 [exec-5] INFO  -               ManageTasks:1037   | MONA: entered ManageTasks.getSelectedTask()
Oct 27 2018 06:47:08 [exec-5] DEBUG -               NgbwSupport:1029   | 046f34e64f8b7dc1015072dc7aa30a1a@ucsd.edu - Error cloning selected task: java.lang.NullPointerException
java.lang.NullPointerException
    at org.ngbw.sdk.database.SourceDocumentRow.<init>(SourceDocumentRow.java:135)
    at org.ngbw.sdk.database.TaskInputSourceDocument.<init>(TaskInputSourceDocument.java:121)
    at org.ngbw.sdk.database.Task.<init>(Task.java:662)
    at org.ngbw.sdk.WorkbenchSession.cloneTask(WorkbenchSession.java:657)
    at org.ngbw.web.actions.CreateTask.cloneTask(CreateTask.java:984)
    at org.ngbw.web.actions.CreateTask.clone(CreateTask.java:130)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

**Comment by monaw** [comment]
tech notes:

- WorkbenchSession.cloneTask creates new Task
- Adding logic in TaskInputSourceDocument to not create a SourceDocumentRow when document's source document id = 0 (this means it is a directory)

**Comment by kenneth59715** [comment]
Actually, refreshFolderDataTabs() in getCurrentAllDataTab() is the one needed for
seeing new globus data items, and it is still there:
    public List<? extends FolderItem> getCurrentAllDataTab()
    {   
        //logger.debug ( "MONA: entered getCurrentAllDataTab()" );
        // added this or else new Globus transfer files won't show up in UI
        // without logout/login
        refreshFolderDataTabs();

So, not sure why it is not showing up now.

**Comment by mcianfrocco** [comment]
Now I'm closing this as I just cloned a task successfully

**Comment by monaw** [comment]
hmm, this "glitch" is worrisome...wonder if something is changing...
are you able to consistently clone a task now?

**Comment by mcianfrocco** [comment]
Hi @monaw & @kenneth59715 - I'm reopening this as I get this error for the 3d refinement task if I clone the task and try to run it.

**Comment by mcianfrocco** [comment]
I believe this is due to the addition of the individual file in the parameters page - where user provided 3D volume file is likely the culprit for the cloning issue.

**Comment by monaw** [comment]
ah that makes sense; what about the 3d initial model tool which doesn't require a volume file?

**Comment by mcianfrocco** [comment]
I think that works - although I'll to check

**Comment by mcianfrocco** [comment]
3D refinement & 3D classification fail for cloning, but 2D classification and 3D initial model are able to be cloned.

**Comment by monaw** [comment]
currently blocked by #150...

**Comment by monaw** [comment]
I just tested 3d classification task cloning (no job submission) and it worked.  now testing another one after job submission...

**Comment by monaw** [comment]
ok 3d refinement task cloning (no job submission) also worked...now testing job submission...

**Comment by monaw** [comment]
@mcianfrocco - ok, i re-cloned the same 3d refine and 3d class tasks after i submitted the job and the cloning were successful.

Please verify...

**Comment by monaw** [comment]
the only "issue" i see is that i have to click on the "Save Task" button twice to save the newly cloned task...

**Comment by mcianfrocco** [comment]
I just tried cloning a currently running task for 3D refinement and I got this error: 

Fri Mar 15 06:37:05 PDT 2019 > INPUTSTAGING : ERROR : NGBW-JOB-RELION_3D_REFINE_COMET-E9C1E6EB230F4F7698B576233A039681 : java.lang.NullPointerException

**Comment by kenneth59715** [comment]
Here's some text from submitJob.log:
...
Mar 15 2019 06:42:35 [read-1] DEBUG -        AsyncProcessRunner:210   [task=890, job=NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424] | Completed :  ssh -i /users/u2/cosmic2-gw/.ssh/id_rsa cosmic2@comet.sdsc.edu source /home/cosmic2/.bash_profile; export WB_JOBID=NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424; cd /projects/cosmic2/gateway/workspace/NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424/; comet_relion_submit.py --account TG-MCB170058 --url '-k https://cosmic2.sdsc.edu:8443/gateway/taskupdate.action?taskId=890\&jh=NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424' -- 'pipeline --bfactor 100 --ctf_res_lim 18 --negative_stain --do_cryolo --cs 2.7 --out_box 0 --tile_frames 5 --apix 1.68 --do_ctf --diameter 110 --movie_binning 1 --out_apix 0 --kev 120 --i "NS_test/mcianfro/vac8-vac17-myo2/19mar05b/Extract/job008/particles.star" '
Mar 15 2019 06:42:35 [read-1] WARN  -      SSHExecProcessWorker:284   [task=890, job=NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424] | Error submitting job: /projects/cosmic2/gateway/workspace/NGBW-JOB-PARTICLE_PICKING_CTF_MOTION-789800872EAA4DED967BF5CDB8EE9424
. Traceback (most recent call last):
  File "/home/cosmic2/COSMIC-CryoEM-Gateway/remote_scripts/comet_relion_submit.py", line 832, in <module>
 movie_align_cmd,ctf_cmd,ctf_sel_cmd,picking_cmd,extraction_cmd,mics_for_picking_symlink,class2d_cmd,cryolo_cp_cmd,DirToSymLink=preparePreprocessingRun(args['commandline'],jobdir)
  File "/home/cosmic2/COSMIC-CryoEM-Gateway/remote_scripts/comet_relion_submit.py", line 277, in preparePreprocessingRun
    for line in open(input_starfile,'r'):
IOError: [Errno 2] No such file or directory: 'mcianfro/vac8-vac17-myo2/19mar05b/Extract/job008/particles.star'
...

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by mcianfrocco** [comment]
that error is from a different problem of mine...

**Comment by kenneth59715** [comment]
Oops, sorry.  This one might be related to finding the source document
and source document id.  I think Mona did something to fix that for task
cloning.  I see debug lines commented out in DefaultToolResource.java.
Is Mona scheduled to work on COSMIC2 today?  If not, I'll uncomment
the debug lines to get more info.

Kenneth

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by monaw** [comment]
@mcianfrocco - is the problem from cloning or from job submission?  The error you posted seems to be from job submission.

i just cloned a 3d refinement task that was running and it was successful...

**Comment by monaw** [comment]
@mcianfrocco - ok I can replicate your error now which comes from actually submitting the newly cloned task; i'll work on this right now...

**Comment by monaw** [comment]
tech note: problem is that task_input_source_documents table's SOURCE_DOCUMENT_ID field is not set on the cloned task

**Comment by kenneth59715** [comment]
Yes, I think that's what I found when I looked at it a while back.
I think globus transfer data items might need a document id
and maybe a hash value.

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by monaw** [comment]
@mcianfrocco - ok **newly** cloned task can now be submitted successfully as jobs, previously cloned task will NOT work because their database entries are missing some information.  i just cloned a 3d refinement task and the job is currently running...

**Comment by monaw** [comment]
@kenneth59715 - the fix was to add the newly created m_sourceDocumentId object to the m_columns object and then it will be saved; anything not in m_columns will not be saved!  i added a note about this in the developer references notes google doc.  i just sent you access through google so you should get an email about that...

**Comment by mcianfrocco** [comment]
I just got an error cloning a 3D refinement: 
Mon Mar 18 19:14:37 PDT 2019 > INPUTSTAGING : ERROR : NGBW-JOB-RELION_3D_REFINE_COMET-6766EF5DDE3B4F70B4B15D14A808218C : java.lang.NullPointerException

**Comment by monaw** [comment]
was the task you cloned FROM an older task created before the fix or new one after fix?

**Comment by mcianfrocco** [comment]
The task was created after your update(3/16/19, 16:25) but the data was uploaded prior to the change. Should I try again from the beginning by uploading new data?

**Comment by monaw** [comment]
hmm, no the data upload should not impact this; what's the name of your 2 tasks (cloned from and to)...let me look into the database...

**Comment by mcianfrocco** [comment]
Clone from: NGBW-JOB-RELION_3D_REFINE_COMET-45950F56FA7D4D2E9835CD08490F663A 
Clone to: NGBW-JOB-RELION_3D_REFINE_COMET-6766EF5DDE3B4F70B4B15D14A808218C

**Comment by monaw** [comment]
just to confirm, NGBW-JOB-RELION_3D_REFINE_COMET-45950F56FA7D4D2E9835CD08490F663A task name is "test" created 3/16 and NGBW-JOB-RELION_3D_REFINE_COMET-6766EF5DDE3B4F70B4B15D14A808218C task name is "test_1" create 3/18?

ok, i think i spot the problem...i commented out my debugging code incorrectly which caused a logic problem in the code; i didn't see the problem because i ran my tests BEFORE i commented out the debugging code, doh!  ok, i just fixed the logic problem so please do another clone and that should run...

**Comment by mcianfrocco** [comment]
awesome - that worked! thanks @monaw ! I'll close the issue for now until it comes up again...

**Comment by monaw** [comment]
haha, hopefully it won't happen again...i'll clean up the code more carefully this time!
